File: generator/build-generator.py
import os
import re
import json
import subprocess


# ============================================================
# recipes.js mit Node auslesen
# ============================================================

node_script = r"""
const fs = require("fs");

let content = fs.readFileSync("recipes.js", "utf8");

content = content.replace(
    "const RECIPES =",
    "global.RECIPES ="
);

eval(content);

console.log(JSON.stringify(global.RECIPES));
"""


result = subprocess.run(
    ["node", "-e", node_script],
    capture_output=True,
    text=True
)


if result.returncode != 0:
    print("FEHLER beim Auslesen von recipes.js:")
    print(result.stderr)
    raise SystemExit(1)


try:
    recipes = json.loads(result.stdout)
except json.JSONDecodeError:
    print("FEHLER: recipes.js konnte nicht als JSON gelesen werden.")
    print(result.stdout)
    raise SystemExit(1)


print()
print("========================================")
print("RECIPES ERFOLGREICH GELADEN")
print(f"{len(recipes)} Rezepte gefunden.")
print("========================================")
print()


# ============================================================
# recipes-Ordner erstellen
# ============================================================

os.makedirs("recipes", exist_ok=True)


# ============================================================
# Slug erstellen
# ============================================================

def create_slug(title):

    slug = str(title).lower()

    slug = (
        slug
        .replace("ä", "ae")
        .replace("ö", "oe")
        .replace("ü", "ue")
        .replace("ß", "ss")
    )

    slug = re.sub(
        r"[^a-z0-9\s-]",
        "",
        slug
    )

    slug = re.sub(
        r"\s+",
        "-",
        slug
    )

    slug = re.sub(
        r"-+",
        "-",
        slug
    )

    return slug.strip("-")


# ============================================================
# [[slug|Text]] in HTML-Link umwandeln
# ============================================================

def replace_recipe_links(text):

    if not isinstance(text, str):
        return text

    pattern = r"\[\[([^|\]]+)\|([^\]]+)\]\]"

    def replace(match):

        slug = match.group(1).strip()
        title = match.group(2).strip()

        return (
            f'<a href="{slug}.html">'
            f'{title}'
            f'</a>'
        )

    return re.sub(
        pattern,
        replace,
        text
    )


# ============================================================
# Alle Rezepte als JSON vorbereiten
# ============================================================

all_recipes_data = json.dumps(
    recipes,
    ensure_ascii=False,
    indent=4
)


# ============================================================
# Template laden
# ============================================================

template_path = "templates/recipe-template.html"


if not os.path.exists(template_path):

    print("FEHLER: Template nicht gefunden:")
    print(template_path)

    raise SystemExit(1)


with open(
    template_path,
    "r",
    encoding="utf-8"
) as file:

    template_original = file.read()


# ============================================================
# Prüfen, ob Platzhalter vorhanden ist
# ============================================================

if "<!-- RECIPE_DATA -->" not in template_original:

    print(
        "FEHLER: <!-- RECIPE_DATA --> "
        "wurde im recipe-template.html nicht gefunden."
    )

    raise SystemExit(1)


# ============================================================
# Jedes Rezept erzeugen
# ============================================================

for recipe in recipes:

    # --------------------------------------------------------
    # Titel prüfen
    # --------------------------------------------------------

    if "title" not in recipe:

        print("FEHLER: Ein Rezept besitzt keinen Titel.")
        print(recipe)

        raise SystemExit(1)


    # --------------------------------------------------------
    # Slug
    # --------------------------------------------------------

    slug = create_slug(
        recipe["title"]
    )


    # --------------------------------------------------------
    # Kopie erstellen
    # --------------------------------------------------------

    recipe_copy = json.loads(
        json.dumps(recipe)
    )


    # --------------------------------------------------------
    # Hauptbild für recipes-Unterordner
    # --------------------------------------------------------

    if "image" in recipe_copy:

        image = recipe_copy["image"]

        if (
            isinstance(image, str)
            and image
            and not image.startswith("../")
            and not image.startswith("http://")
            and not image.startswith("https://")
        ):

            recipe_copy["image"] = (
                "../" + image
            )


    # --------------------------------------------------------
    # Schrittbilder für recipes-Unterordner
    # --------------------------------------------------------

    for step in recipe_copy.get(
        "steps",
        []
    ):

        if (
            isinstance(step, dict)
            and "images" in step
        ):

            new_images = []

            for img in step["images"]:

                if (
                    isinstance(img, str)
                    and img
                    and not img.startswith("../")
                    and not img.startswith("http://")
                    and not img.startswith("https://")
                ):

                    new_images.append(
                        "../" + img
                    )

                else:

                    new_images.append(img)

            step["images"] = new_images


    # --------------------------------------------------------
    # Zutatenlinks
    # --------------------------------------------------------

    ingredients = recipe_copy.get(
        "ingredients"
    )


    if isinstance(
        ingredients,
        dict
    ):

        for section, items in ingredients.items():

            if isinstance(items, list):

                recipe_copy["ingredients"][section] = [
                    replace_recipe_links(item)
                    for item in items
                ]


    elif isinstance(
        ingredients,
        list
    ):

        recipe_copy["ingredients"] = [
            replace_recipe_links(item)
            for item in ingredients
        ]


    # --------------------------------------------------------
    # Tipps verlinken
    # --------------------------------------------------------

    if "tips" in recipe_copy:

        if isinstance(
            recipe_copy["tips"],
            list
        ):

            recipe_copy["tips"] = [
                replace_recipe_links(tip)
                for tip in recipe_copy["tips"]
            ]


    # --------------------------------------------------------
    # Dateiname
    # --------------------------------------------------------

    filename = (
        f"recipes/{slug}.html"
    )


    print("========================================")
    print("Verarbeite Rezept:")
    print(recipe["title"])
    print("Datei:")
    print(filename)
    print("========================================")


    # --------------------------------------------------------
    # Template zurücksetzen
    # --------------------------------------------------------

    template = template_original


    # --------------------------------------------------------
    # Aktuelles Rezept als JSON
    # --------------------------------------------------------

    recipe_data = json.dumps(
        recipe_copy,
        ensure_ascii=False,
        indent=4
    )


    # --------------------------------------------------------
    # RECIPE_DATA + RECIPES einsetzen
    # --------------------------------------------------------

    replacement = f"""
<script>
const RECIPE_DATA = {recipe_data};
const RECIPES = {all_recipes_data};
</script>
"""


    template = template.replace(
        "<!-- RECIPE_DATA -->",
        replacement,
        1
    )


    # --------------------------------------------------------
    # HTML speichern
    # --------------------------------------------------------

    with open(
        filename,
        "w",
        encoding="utf-8"
    ) as file:

        file.write(template)


    # --------------------------------------------------------
    # Prüfen
    # --------------------------------------------------------

    if os.path.exists(filename):

        print("ERSTELLT:")
        print(
            os.path.abspath(filename)
        )
        print()

    else:

        print("FEHLER: Datei wurde nicht erstellt:")
        print(
            os.path.abspath(filename)
        )

        raise SystemExit(1)


# ============================================================
# SITEMAP ERSTELLEN
# ============================================================

sitemap_lines = [
    '<?xml version="1.0" encoding="UTF-8"?>',
    '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
    "",
    "    <url>",
    "        <loc>https://kitchenbreeze.github.io/foodblog/</loc>",
    "    </url>",
    ""
]


# ============================================================
# Rezeptseiten zur Sitemap hinzufügen
# ============================================================

for recipe in recipes:

    slug = create_slug(
        recipe["title"]
    )


    sitemap_lines.extend([
        "    <url>",
        (
            "        <loc>"
            f"https://kitchenbreeze.github.io/foodblog/recipes/{slug}.html"
            "</loc>"
        ),
        "    </url>",
        ""
    ])


# ============================================================
# Sitemap schließen
# ============================================================

sitemap_lines.append(
    "</urlset>"
)


# ============================================================
# sitemap.xml speichern
# ============================================================

with open(
    "sitemap.xml",
    "w",
    encoding="utf-8"
) as file:

    file.write(
        "\n".join(
            sitemap_lines
        )
    )


# ============================================================
# FERTIG
# ============================================================

print()
print("========================================")
print("SITEMAP ERSTELLT")
print("sitemap.xml")
print("========================================")
print()

print("========================================")
print("GENERATOR ERFOLGREICH ABGESCHLOSSEN")
print(f"{len(recipes)} Rezepte verarbeitet.")
print("========================================")

File: generator/build-generator.js
const fs = require("fs");
const path = require("path");

// ============================================================
// EINSTELLUNGEN
// ============================================================

const RECIPES_JS = "recipes.js";
const TEMPLATE_PATH = path.join("templates", "recipe-template.html");
const OUTPUT_DIR = "recipes";
const SITEMAP_PATH = "sitemap.xml";

const BASE_URL = "https://kitchenbreeze.github.io/foodblog";

// ============================================================
// START
// ============================================================

console.log();
console.log("========================================");
console.log("       KITCHENBREEZE GENERATOR");
console.log("========================================");
console.log();

// ============================================================
// DATEIEN PRÜFEN
// ============================================================

if (!fs.existsSync(RECIPES_JS)) {
    console.error("FEHLER:");
    console.error(`${RECIPES_JS} wurde nicht gefunden.`);
    process.exit(1);
}

if (!fs.existsSync(TEMPLATE_PATH)) {
    console.error("FEHLER:");
    console.error("Template nicht gefunden:");
    console.error(TEMPLATE_PATH);
    process.exit(1);
}

// ============================================================
// RECIPES.JS LADEN
// ============================================================

let recipesContent = fs.readFileSync(
    RECIPES_JS,
    "utf8"
);

// ============================================================
// RECIPES.JS AUSLESEN
// ============================================================

const nodeScript = `
const fs = require("fs");

let content = fs.readFileSync(${JSON.stringify(RECIPES_JS)}, "utf8");

content = content.replace(
    /const\\\\s+RECIPES\\\\s*=/,
    "global.RECIPES ="
);

eval(content);

console.log(JSON.stringify(global.RECIPES));
`;

console.log("Lese recipes.js ...");
console.log();

const result = require("child_process").spawnSync(
    process.execPath,
    ["-e", nodeScript],
    {
        encoding: "utf8"
    }
);

if (result.error) {
    console.error("FEHLER beim Starten von Node.js:");
    console.error(result.error.message);
    process.exit(1);
}

if (result.status !== 0) {
    console.error("FEHLER beim Auslesen von recipes.js:");
    console.error();
    console.error(result.stderr);
    process.exit(1);
}

let recipes;

try {
    recipes = JSON.parse(result.stdout);
} catch (error) {
    console.error("FEHLER:");
    console.error("recipes.js konnte nicht gelesen werden.");
    console.error();
    console.error(result.stdout);
    process.exit(1);
}

if (!Array.isArray(recipes)) {
    console.error("FEHLER:");
    console.error("RECIPES ist kein Array.");
    process.exit(1);
}

console.log("RECIPES.JS ERFOLGREICH GELADEN");
console.log(`${recipes.length} Rezepte gefunden.`);
console.log();

// ============================================================
// OUTPUT-ORDNER ERSTELLEN
// ============================================================

fs.mkdirSync(
    OUTPUT_DIR,
    {
        recursive: true
    }
);

// ============================================================
// SLUG ERSTELLEN
// ============================================================

function createSlug(title) {

    let slug = String(title)
        .toLowerCase()
        .trim();

    slug = slug
        .replace(/ä/g, "ae")
        .replace(/ö/g, "oe")
        .replace(/ü/g, "ue")
        .replace(/ß/g, "ss");

    slug = slug.replace(
        /[^a-z0-9\s-]/g,
        ""
    );

    slug = slug.replace(
        /\s+/g,
        "-"
    );

    slug = slug.replace(
        /-+/g,
        "-"
    );

    return slug.replace(
        /^-+|-+$/g,
        ""
    );
}

// ============================================================
// BILDPFAD FÜR REZEPTSEITEN
//
// Rezeptseite:
//
// recipes/pasta.html
//
// Bild:
//
// images/pasta.jpg
//
// Ergebnis:
//
// ../images/pasta.jpg
// ============================================================

function recipePageImagePath(imagePath) {

    if (typeof imagePath !== "string") {
        return imagePath;
    }

    let value = imagePath.trim();

    if (!value) {
        return value;
    }

    // Externe Bilder nicht verändern
    if (
        value.startsWith("http://") ||
        value.startsWith("https://")
    ) {
        return value;
    }

    // Data-URLs nicht verändern
    if (value.startsWith("data:")) {
        return value;
    }

    // Absolute Pfade nicht verändern
    if (value.startsWith("/")) {
        return value;
    }

    // Bereits korrekter Pfad
    if (value.startsWith("../")) {
        return value;
    }

    // ./ entfernen
    if (value.startsWith("./")) {
        value = value.substring(2);
    }

    // Führenden Slash entfernen
    value = value.replace(/^\/+/, "");

    return "../" + value;
}

// ============================================================
// [[slug|Text]] IN HTML-LINK UMWANDELN
// ============================================================

function replaceRecipeLinks(text) {

    if (typeof text !== "string") {
        return text;
    }

    const pattern =
        /\[\[([^|\]]+)\|([^\]]+)\]\]/g;

    return text.replace(
        pattern,
        function (
            match,
            slug,
            title
        ) {

            slug = slug.trim();
            title = title.trim();

            return (
                `<a href="../recipes/${slug}.html">` +
                `${title}` +
                `</a>`
            );
        }
    );
}

// ============================================================
// REZEPT FÜR REZEPTSEITE VORBEREITEN
// ============================================================

function prepareRecipeForPage(recipe) {

    // Tiefe Kopie
    const recipeCopy = JSON.parse(
        JSON.stringify(recipe)
    );

    // ========================================================
    // HAUPTBILD
    // ========================================================

    if (
        Object.prototype.hasOwnProperty.call(
            recipeCopy,
            "image"
        )
    ) {

        recipeCopy.image =
            recipePageImagePath(
                recipeCopy.image
            );
    }

    // ========================================================
    // SCHRITTBILDER
    // ========================================================

    if (Array.isArray(recipeCopy.steps)) {

        recipeCopy.steps.forEach(
            function (step) {

                if (
                    !step ||
                    typeof step !== "object"
                ) {
                    return;
                }

                if (
                    !Array.isArray(step.images)
                ) {
                    return;
                }

                step.images =
                    step.images.map(
                        function (image) {

                            return recipePageImagePath(
                                image
                            );
                        }
                    );
            }
        );
    }

    // ========================================================
    // ZUTATEN
    // ========================================================

    const ingredients =
        recipeCopy.ingredients;

    if (
        ingredients &&
        typeof ingredients === "object" &&
        !Array.isArray(ingredients)
    ) {

        Object.keys(ingredients).forEach(
            function (section) {

                if (
                    !Array.isArray(
                        ingredients[section]
                    )
                ) {
                    return;
                }

                ingredients[section] =
                    ingredients[section].map(
                        replaceRecipeLinks
                    );
            }
        );

    } else if (
        Array.isArray(ingredients)
    ) {

        recipeCopy.ingredients =
            ingredients.map(
                replaceRecipeLinks
            );
    }

    // ========================================================
    // TIPPS
    // ========================================================

    if (
        Array.isArray(recipeCopy.tips)
    ) {

        recipeCopy.tips =
            recipeCopy.tips.map(
                replaceRecipeLinks
            );
    }

    return recipeCopy;
}

// ============================================================
// ALLE REZEPTE FÜR RELATED RECIPES VORBEREITEN
// ============================================================

const recipesForRecipePages =
    recipes.map(
        prepareRecipeForPage
    );

const allRecipesData =
    JSON.stringify(
        recipesForRecipePages,
        null,
        4
    );

// ============================================================
// TEMPLATE LADEN
// ============================================================

const templateOriginal =
    fs.readFileSync(
        TEMPLATE_PATH,
        "utf8"
    );

// ============================================================
// PLACEHOLDER
// ============================================================

const placeholder =
    "<!-- RECIPE_DATA -->";

if (
    !templateOriginal.includes(
        placeholder
    )
) {

    console.error("FEHLER:");
    console.error(
        "<!-- RECIPE_DATA --> wurde nicht gefunden."
    );

    process.exit(1);
}

// ============================================================
// REZEPTSEITEN ERSTELLEN
// ============================================================

let createdCount = 0;

for (const recipe of recipes) {

    // ========================================================
    // TITEL PRÜFEN
    // ========================================================

    if (
        !recipe ||
        typeof recipe !== "object"
    ) {

        console.error(
            "FEHLER: Ungültiges Rezept gefunden."
        );

        process.exit(1);
    }

    if (
        typeof recipe.title !== "string" ||
        !recipe.title.trim()
    ) {

        console.error(
            "FEHLER: Ein Rezept besitzt keinen Titel."
        );

        console.error(recipe);

        process.exit(1);
    }

    const title =
        recipe.title.trim();

    // ========================================================
    // SLUG
    // ========================================================

    const slug =
        createSlug(title);

    if (!slug) {

        console.error("FEHLER:");
        console.error(
            `Kein gültiger Dateiname für: ${title}`
        );

        process.exit(1);
    }

    // ========================================================
    // REZEPTDATEN VORBEREITEN
    // ========================================================

    const recipeCopy =
        prepareRecipeForPage(
            recipe
        );

    // ========================================================
    // DATEINAME
    // ========================================================

    const filename =
        path.join(
            OUTPUT_DIR,
            `${slug}.html`
        );

    console.log("----------------------------------------");
    console.log("Rezept:");
    console.log(title);
    console.log();
    console.log("Datei:");
    console.log(filename);

    // ========================================================
    // TEMPLATE ZURÜCKSETZEN
    // ========================================================

    let template =
        templateOriginal;

    // ========================================================
    // RECIPE_DATA
    // ========================================================

    const recipeData =
        JSON.stringify(
            recipeCopy,
            null,
            4
        );

    // ========================================================
    // JAVASCRIPT DATEN EINSETZEN
    // ========================================================

    const replacement = `
<script>
const RECIPE_DATA = ${recipeData};
const RECIPES = ${allRecipesData};
</script>
`;

    template =
        template.replace(
            placeholder,
            replacement
        );

    // ========================================================
    // HTML SCHREIBEN
    // ========================================================

    fs.writeFileSync(
        filename,
        template,
        "utf8"
    );

    // ========================================================
    // PRÜFEN
    // ========================================================

    if (
        !fs.existsSync(filename)
    ) {

        console.error();
        console.error("FEHLER:");
        console.error(
            "Datei wurde nicht erstellt:"
        );
        console.error(
            path.resolve(filename)
        );

        process.exit(1);
    }

    createdCount++;

    console.log("OK - ERSTELLT:");
    console.log(
        path.resolve(filename)
    );
    console.log();
}

// ============================================================
// SITEMAP ERSTELLEN
// ============================================================

const sitemapLines = [];

sitemapLines.push(
    '<?xml version="1.0" encoding="UTF-8"?>'
);

sitemapLines.push(
    `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`
);

sitemapLines.push("");

sitemapLines.push("    <url>");

sitemapLines.push(
    `        <loc>${BASE_URL}/</loc>`
);

sitemapLines.push("    </url>");

sitemapLines.push("");

// ============================================================
// REZEPTSEITEN IN SITEMAP
// ============================================================

for (const recipe of recipes) {

    const slug =
        createSlug(
            recipe.title
        );

    sitemapLines.push(
        "    <url>"
    );

    sitemapLines.push(
        `        <loc>${BASE_URL}/recipes/${slug}.html</loc>`
    );

    sitemapLines.push(
        "    </url>"
    );

    sitemapLines.push("");
}

// ============================================================
// SITEMAP SCHLIESSEN
// ============================================================

sitemapLines.push(
    "</urlset>"
);

// ============================================================
// SITEMAP SPEICHERN
// ============================================================

fs.writeFileSync(
    SITEMAP_PATH,
    sitemapLines.join("\n"),
    "utf8"
);

// ============================================================
// FERTIG
// ============================================================

console.log();
console.log("========================================");
console.log("   GENERATOR ERFOLGREICH ABGESCHLOSSEN");
console.log("========================================");
console.log();

console.log(
    `${createdCount} Rezeptseiten erstellt.`
);

console.log();

console.log(
    `Sitemap erstellt: ${SITEMAP_PATH}`
);

console.log();

console.log(
    "Bildpfade wurden für recipes/ angepasst."
);

console.log(
    "Hauptbilder und Schrittbilder berücksichtigt."
);

console.log();

console.log("========================================");


